### ISSUE TITLE
parallel process can take up too much memory

### ISSUE THREAD

**Issue by hxin** [sidbdri/cookiecutter-de_analysis_skeleton#197]
in SnowParam setting, each work's memory is independent of the host. For the jobs to run in work, objects need to be passed/copied to the worker from the host.

For example, below will copy the x/y/z object into each worker's memory.
```
bplapply(job_strings,
         'test_function',
         x=ep(c(1:1e6),1),
         y=rep(c(1:1e6),1),
         z=rep(c(1:1e6),1),
         BPPARAM = para
)
```

This can potentially cause when using a large number of cores and when passing large objects. For example, the __dds__ object passed to __run_gene_set_analysis__ function can be 5+GB, when using 32 cores, this itself takes up 160GB memory. It is not only time comsuming but also might cause memory issues.
https://github.com/sidbdri/cookiecutter-de_analysis_skeleton/blob/4b8f6aa77a45814db1c036fb385b031313230784/%7B%7Bcookiecutter.project_name%7D%7D/diff_expr.R#L265

**Comment by hxin** [comment]
Some test code:
```
job_strings=c(1:2)
para <- call('SnowParam', workers = 2, stop.on.error = TRUE, jobname = 'test', progressbar = T,
             log = TRUE, threshold = "DEBUG", logdir = file.path('/tmp'), tasks = length(job_strings)) %>% eval
bpstart(para)
test_function<-function(x,...){
  object_lists <- sapply(list(...),function(x){prettyunits::pretty_bytes(as.numeric(pryr::object_size(x)))})
  total_object_size<-pryr::object_size(list(...))
  
  futile.logger::flog.info('In shared memory mode, objects are copied to each worker in memory.')
  futile.logger::flog.info('See below of the size of objects:')
  print(object_lists)
  
  futile.logger::flog.info('Total size of Objects passing to worker:')
  print(total_object_size)
  
  futile.logger::flog.info('To avoid using a lot of memory when pasing large objects to worker, we limit the number of worker base on the objects size')
  ALLOW_MEMORY=1024^2
  if(as.numeric(total_object_size) > ALLOW_MEMORY)
    message('we need to reduce worker numbers')
  else
    message('looks fine')
}
bplapply(job_strings,
         'test_function',
         y=rep(c(1:1e6),1),
         z=rep(c(1:1e6),1),
         BPPARAM = para
)


bpstop(para)
```

**Comment by hxin** [comment]
More on this issue:

It seems that some function such as __sapply__, when called, memory allocation and garbage is not collected instantaneously.

So if 
`sapply(list(a=rep(c(1:1e6),200)),function(x){pryr::object_size(x)})`
called multiple times, you can see how top mem usage increase.

A manual __gc()__ will free up the memory.

**Comment by hxin** [comment]
Wrote the below code to check the objects passed into the worker but this code it self seems to trigger a duplication of the objects. I log the code here for furture reference and leave it out of our pipeline for now.
```
if (parallelParam == 'SnowParam') {
      object_lists <- sapply(list(...),function(x){prettyunits::pretty_bytes(as.numeric(pryr::object_size(x)))})
      total_object_size<-pryr::object_size(list(...))
      
      message('In SnowParam shared memory mode, objects are copied to each worker in memory.')
      message('See below of the size of objects:')
      print(object_lists)
      
      message('Total size of Objects passing to worker:')
      print(total_object_size)
      
      message('Total memory estimate size of workers:')
      print(total_object_size * nc)
      
      allow_memory=am*1024^3
      message('Maximun memory allowed: ',paste0(am,' Gb'))
      
      if(as.numeric(total_object_size)*nc > allow_memory){
        message('To avoid using a lot of memory when pasing large objects to worker, we limit the number of worker base on the objects size')
        nc=as.integer((allow_memory/total_object_size))
        message('Reduce the number of worker to:',nc)
      }
    }
```

**Comment by hxin** [comment]
The memory useage seems to be larger than expect and it seems to be rooted to the problem that the objects passed to the workers are duplicated. For example, 5GB of objects seems to be copied more than once which ends up using 10GB, 15GB memory.

I tried a few approaches but cant seems to solve this issue. 

We decided for now to try optimize the objects passed to the worker, or to limit the total number of worker.
